### PR TITLE
Use Worker to serialize Notebook

### PR DESCRIPTION
**Pull request by nojaf**
Related to #214040

Hi @amunger, I tried running the `JSON.stringify` call on a Worker thread. It works on my machine and seems to improve the situation, but it's hard to say if the issue is completely resolved. Any thoughts?

### CODE REVIEW

**Comment by nojaf** [comment]
@microsoft-github-policy-service agree company=G-Research

**Comment by amunger** [comment]
nice! I should be able to give this a try at the end of the week.
We would also want to put this behind a setting so we can roll it out with an experiment if it works as expected.

**Comment by nojaf** [comment]
@amunger, I'm not entirely sure about the risks of using a worker for this in Node or how stable it is. It likely depends on whether this practice is already being used elsewhere in the code. If you feel a setting is necessary, I’m fine with it. However, it might be safe enough to proceed without one.

**Comment by amunger** [comment]
here's a sample of the perf diff on a 400mb notebook - it reduces the time that the EH is locked up by ~24%, that big block goes from 2.93s -> 2.23s.
image

It seems like more of the work could be pushed to the worker to reduce this further, `sortObjectPropertiesRecursively` would be pretty straight forward, and `createJupyterCellFromNotebookCell` would be trickier since it's using types from `vscode`.

If possible, we should also follow the pattern of Worker usage from the other built-in extensions and avoid the node imports e.g. the [css extension](https://github.com/microsoft/vscode/blob/f0417069c62e20f3667506f4b7e53ca0004b4e3e/extensions/css-language-features/client/src/node/cssClientMain.ts#L19)
Otherwise, we need to ensure that we only use this in desktop, since a web based EH doesn't have access to node.

**Comment by amunger** [comment]
also, make sure to get the latest since we just had a big change go in to switch over to ESM imports

**Comment by nojaf** [comment]
Thanks a bunch for this feedback. Will circle back to this on Monday. Enjoy the weekend, cheers!

**Comment by nojaf** [comment]
Hi @amunger  

> If possible, we should also follow the pattern of Worker usage from the other built-in extensions and avoid the node imports e.g. the [css extension](https://github.com/microsoft/vscode/blob/f0417069c62e20f3667506f4b7e53ca0004b4e3e/extensions/css-language-features/client/src/node/cssClientMain.ts#L19)

I'm a little unsure how to proceed there, the client/server thing seems like overkill compared to the Worker thread. I'm checking for Node right now and using dynamic imports. Let me know if that works for you or if you would prefer the client/server approach.

> It seems like more of the work could be pushed to the worker to reduce this further, sortObjectPropertiesRecursively would be pretty straight forward,

I actually tried that but got:

```
Failed to save 'smaller.ipynb': Cannot find module 'vscode' Require stack: - c:\Users\nojaf\Projects\vscode\extensions\ipynb\out\serializers.js - c:\Users\nojaf\Projects\vscode\extensions\ipynb\out\notebookSerializerWorker.js
```

the JavaScript is something like

```
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const node_worker_threads_1 = require("node:worker_threads");
const serializers_1 = require("./serializers");
if (node_worker_threads_1.parentPort) {
    const { notebookContent, indentAmount } = node_worker_threads_1.workerData;
    const json = JSON.stringify((0, serializers_1.sortObjectPropertiesRecursively)(notebookContent), undefined, indentAmount) + '\n';
    node_worker_threads_1.parentPort.postMessage(json);
}
//# sourceMappingURL=notebookSerializerWorker.js.map
```

and I think the `require("./serializers")` loads the entire file which somewhere contains `vscode` which is not known in the worker perhaps?

**Comment by amunger** [comment]
> Failed to save 'smaller.ipynb': Cannot find module 'vscode' Require stack

Yeah that's what I would expect - the imports get pretty deep into the vscode API for that function, so we can just leave that out for now.

Limiting to node is fine for now, we'll just have to see if the CI tests have any issues with importing like that.

there are some conflicting files to resolve at this point though.

**Comment by nojaf** [comment]
Alright, I believe this is ready for review @amunger.

**Review by amunger — DISMISSED** [review]

File: extensions/ipynb/src/notebookSerializer.ts
Comment (rebornix) [inline extensions/ipynb/src/notebookSerializer.ts:84]: this will break once the extension is bundled. We need to copy the worker file to dist. Similar to https://github.com/nojaf/vscode/blob/12abcccc8147cb30d5004fb41366b96670bdbf6c/extensions/markdown-language-features/extension.webpack.config.js#L21-L31

File: extensions/ipynb/package.nls.json
Comment (rebornix) [inline extensions/ipynb/package.nls.json:5]: since it only works in node extension host, let's call that out explicitly.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (rebornix) [inline extensions/ipynb/src/notebookSerializer.ts:109]: it seems we dont' append the additional `\n` anymore?

**Review by rebornix — CHANGES_REQUESTED** [review]
It looks great, thank you for your help!

I left a few comments and they might need to be resolved before we merge the PR otherwise it might function as expected.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (nojaf) [reply extensions/ipynb/src/notebookSerializer.ts:84]: Good point, I tried to add the copy plugin. But I'm not sure how I can trigger the bundling?
Is there a `yarn run` script that will bundle the extension, so I can test this?

File: extensions/ipynb/package.nls.json
Comment (nojaf) [reply extensions/ipynb/package.nls.json]: I mentioned this is for desktop, would that be a good description? I was unsure how to phrase it.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (nojaf) [reply extensions/ipynb/src/notebookSerializer.ts]: Yes, good spot, thanks!

**Comment by nojaf** [comment]
Thanks for the reviews @amunger and @rebornix!

File: extensions/ipynb/package.nls.json
Comment (amunger) [inline extensions/ipynb/package.nls.json:5]: It will actually still work in web if you have a remote ssh connection.

perhaps:

```suggestion
	"ipynb.experimental.serialization": "Experimental feature to serialize the Jupyter notebook in a worker thread. Not supported when the Extension host is running as a web worker.",
```

File: extensions/ipynb/src/notebookSerializer.ts
Comment (DonJayamanne) [inline extensions/ipynb/src/notebookSerializer.ts:85]: @rebornix I think we might want to send the bytes using a similar technique like in `stringifyJsonWithBufferRefs`
Also workers have better perf when sending bytes across as opposed to strings (e.g. images and the like).
I think the larger the outputs get, and more complex (different tyeps of mime types), then the benefits would out weight not using it (after all we're using it in rpc layer).

Once we do that we can send the `vscode.NotebookData` object across directly and just import the types in worker and move 90% of the code from this file into a common place so it can be used in both workers as well as extension host context.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (DonJayamanne) [inline extensions/ipynb/src/notebookSerializer.ts:79]: Why only node.js? I think we can get this into web as well.
@rebornix or we could perhaps get this PR in, and then seaprately add support for the other items (web, stringifyJsonWithBufferRefs, etc)
Else this PR might take a while to get in.
I've tried a few things at my end, and am happy to add them to this PR or add them after this PR is merged.
Let discuss offline.

Comment (DonJayamanne) [reply extensions/ipynb/src/notebookSerializer.ts:79]: @rebornix another idea was using the `extension.affninity` setting, to get this ext into a different extension host (however Jupyter relies on some functionality from this), we could split this extension, but not idea.
Just thought I'd mention it (that way it would unblock ext host and free for kernel execution )

File: extensions/ipynb/src/notebookSerializer.ts
Comment (rebornix) [reply extensions/ipynb/src/notebookSerializer.ts:84]: sorry for misleading you here, local build and product bundling are behaving differently. I pushed a commit to fix this and will give this a test.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (amunger) [reply extensions/ipynb/src/notebookSerializer.ts:79]: we can add web support later

File: extensions/ipynb/extension.webpack.config.js
Comment (DonJayamanne) [inline extensions/ipynb/extension.webpack.config.js:16]: @rebornix I do not think this will work,
We will need to ensure we have 2 builds, one for browser and one for web,
trying to import `worker_threads` when building for web could fall over in build, thats what I think.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (DonJayamanne) [inline extensions/ipynb/src/notebookSerializer.ts:100]: @nojaf Thanks for the great work, 
I've made one minor change (looks like a lot though),
basically I'm sending the entire `vscode.NotebookData` to the worker, and processing everything there.
That object can be safely serialized and sent to the worker

This way, all of the cup intensive work is done in worker, including constructing the JSON, serialization to bytes and the sorting of object keys.

& the other change, it to use a single worker as opposed to starting one each time,
this saves a lot more time

File: extensions/ipynb/package.json
Comment (DonJayamanne) [inline extensions/ipynb/package.json:26]: This way we have a full proof way to determine whether the runtime context is browser or node.js
See `ipynbMain.ts` file. (passing extra argument `isBrowser`)

File: extensions/ipynb/extension.webpack.config.js
Comment (DonJayamanne) [inline extensions/ipynb/extension.webpack.config.js:17]: We need this notebookSerializerWorker entry as we need to compile a new output, just copying this will not work (as this file as additional imports).

File: extensions/ipynb/src/notebookSerializer.ts
Comment (DonJayamanne) [inline extensions/ipynb/src/notebookSerializer.ts:156]: Same technique used in other built in extensions.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (nojaf) [reply extensions/ipynb/src/notebookSerializer.ts:100]: Hi @DonJayamanne! 

Thank you so much for refining this PR! It’s such a significant improvement over my original code, and I genuinely appreciate the effort you put into it—this would have taken me quite a while to figure out on my own.

Everything is still functioning perfectly on my end, and I'm excited about moving forward with this!

Thanks again! If there’s anything else I can do to help, just give me a shout!

**Comment by DonJayamanne** [comment]
@amunger @rebornix , please can you review this.

File: extensions/ipynb/src/notebookSerializer.ts
Comment (nojaf) [inline extensions/ipynb/src/notebookSerializer.ts:136]: Nitpick, should this be renamed to `serializeViaWorker` instead of `serializeViaWorker2`?

File: extensions/ipynb/src/notebookSerializer.ts
Comment (nojaf) [inline extensions/ipynb/src/notebookSerializer.ts:143]: Can this be removed or do you want to keep this in @DonJayamanne?

**Review by amunger — APPROVED** [review]

**Comment by nojaf** [comment]
Hello @rebornix, would you mind reviewing this PR as well?
Is this good to go? Let me know if this needs any more tweaks, happy to address these.

**Comment by rebornix** [comment]
It looks good to me. Before we go ahead and merge, @DonJayamanne did you get a chance to run a build on CI so we can verify if it works as expected?

**Comment by DonJayamanne** [comment]
> did you get a chance to run a build on CI so we can verify if it works as expected?

Yes.

**Review by rebornix — APPROVED** [review]

**Pull request merged**
